File: examples/04-pki-and-x509/cpp/main.cpp
// =============================================================================
// Applied PQC Lab - PQC X.509 PKI End-to-End Encryption in C++20
// (ML-DSA-65 Root CA + ML-KEM-768 Leaf Cert + AES-256-GCM)
// =============================================================================

#include <iostream>
#include <vector>
#include <string>
#include <memory>
#include <stdexcept>
#include <cassert>
#include <openssl/x509.h>
#include <openssl/x509_vfy.h>
#include <openssl/pem.h>
#include <openssl/evp.h>

struct X509Deleter { void operator()(X509* p) const { if (p) X509_free(p); } };
struct X509StoreDeleter { void operator()(X509_STORE* p) const { if (p) X509_STORE_free(p); } };
struct X509StoreCtxDeleter { void operator()(X509_STORE_CTX* p) const { if (p) X509_STORE_CTX_free(p); } };
struct EvpPkeyDeleter { void operator()(EVP_PKEY* p) const { if (p) EVP_PKEY_free(p); } };
struct EvpPkeyCtxDeleter { void operator()(EVP_PKEY_CTX* p) const { if (p) EVP_PKEY_CTX_free(p); } };
struct EvpCipherCtxDeleter { void operator()(EVP_CIPHER_CTX* p) const { if (p) EVP_CIPHER_CTX_free(p); } };

using ScopedX509 = std::unique_ptr<X509, X509Deleter>;
using ScopedX509Store = std::unique_ptr<X509_STORE, X509StoreDeleter>;
using ScopedX509StoreCtx = std::unique_ptr<X509_STORE_CTX, X509StoreCtxDeleter>;
using ScopedPKEY = std::unique_ptr<EVP_PKEY, EvpPkeyDeleter>;
using ScopedPKEY_CTX = std::unique_ptr<EVP_PKEY_CTX, EvpPkeyCtxDeleter>;
using ScopedCipherCtx = std::unique_ptr<EVP_CIPHER_CTX, EvpCipherCtxDeleter>;

std::pair<std::vector<uint8_t>, std::vector<uint8_t>> aes_gcm_encrypt(
    const std::vector<uint8_t>& key, const std::vector<uint8_t>& iv,
    const std::string& plaintext, const std::string& aad)
{
    ScopedCipherCtx ctx(EVP_CIPHER_CTX_new());
    EVP_EncryptInit_ex(ctx.get(), EVP_aes_256_gcm(), nullptr, nullptr, nullptr);
    EVP_CIPHER_CTX_ctrl(ctx.get(), EVP_CTRL_GCM_SET_IVLEN, iv.size(), nullptr);
    EVP_EncryptInit_ex(ctx.get(), nullptr, nullptr, key.data(), iv.data());

    int outlen = 0;
    if (!aad.empty()) {
        EVP_EncryptUpdate(ctx.get(), nullptr, &outlen, reinterpret_cast<const uint8_t*>(aad.data()), aad.size());
    }
    std::vector<uint8_t> ct(plaintext.size() + 16);
    EVP_EncryptUpdate(ctx.get(), ct.data(), &outlen, reinterpret_cast<const uint8_t*>(plaintext.data()), plaintext.size());
    int total_len = outlen;
    int final_len = 0;
    EVP_EncryptFinal_ex(ctx.get(), ct.data() + total_len, &final_len);
    total_len += final_len;
    ct.resize(total_len);

    std::vector<uint8_t> tag(16);
    EVP_CIPHER_CTX_ctrl(ctx.get(), EVP_CTRL_GCM_GET_TAG, 16, tag.data());
    return {ct, tag};
}

std::string aes_gcm_decrypt(
    const std::vector<uint8_t>& key, const std::vector<uint8_t>& iv,
    const std::vector<uint8_t>& ct, const std::vector<uint8_t>& tag, const std::string& aad)
{
    ScopedCipherCtx ctx(EVP_CIPHER_CTX_new());
    EVP_DecryptInit_ex(ctx.get(), EVP_aes_256_gcm(), nullptr, nullptr, nullptr);
    EVP_CIPHER_CTX_ctrl(ctx.get(), EVP_CTRL_GCM_SET_IVLEN, iv.size(), nullptr);
    EVP_DecryptInit_ex(ctx.get(), nullptr, nullptr, key.data(), iv.data());

    int outlen = 0;
    if (!aad.empty()) {
        EVP_DecryptUpdate(ctx.get(), nullptr, &outlen, reinterpret_cast<const uint8_t*>(aad.data()), aad.size());
    }
    std::vector<uint8_t> pt(ct.size() + 16);
    EVP_DecryptUpdate(ctx.get(), pt.data(), &outlen, ct.data(), ct.size());
    int total_len = outlen;
    EVP_CIPHER_CTX_ctrl(ctx.get(), EVP_CTRL_GCM_SET_TAG, 16, const_cast<uint8_t*>(tag.data()));
    int final_len = 0;
    int ret = EVP_DecryptFinal_ex(ctx.get(), pt.data() + total_len, &final_len);
    if (ret <= 0) throw std::runtime_error("AES-GCM Authentication Failed!");
    total_len += final_len;
    return std::string(reinterpret_cast<char*>(pt.data()), total_len);
}

int main(int argc, char* argv[]) {
    std::string pki_dir = (argc > 1) ? argv[1] : "/tmp/pqc_pki_out";

    std::cout << "======================================================\n";
    std::cout << " [Applied PQC Lab] PQC X.509 PKI End-to-End in C++20\n";
    std::cout << " (ML-DSA-65 Root CA + ML-KEM-768 Leaf + AES-256-GCM)\n";
    std::cout << "======================================================\n";

    try {
        // 1. Load Root CA and Leaf Certs
        std::cout << "[+] Step 1: Loading Root CA (ML-DSA-65) and Receiver Cert (ML-KEM-768)...\n";
        FILE* ca_f = fopen((pki_dir + "/ca.crt").c_str(), "r");
        if (!ca_f) throw std::runtime_error("Failed to open ca.crt");
        ScopedX509 ca_cert(PEM_read_X509(ca_f, nullptr, nullptr, nullptr));
        fclose(ca_f);

        FILE* leaf_f = fopen((pki_dir + "/receiver.crt").c_str(), "r");
        if (!leaf_f) throw std::runtime_error("Failed to open receiver.crt");
        ScopedX509 leaf_cert(PEM_read_X509(leaf_f, nullptr, nullptr, nullptr));
        fclose(leaf_f);

        // 2. Sender verifies Certificate Chain
        std::cout << "[+] Step 2: Sender verifying Receiver X.509 Certificate Chain...\n";
        ScopedX509Store store(X509_STORE_new());
        X509_STORE_add_cert(store.get(), ca_cert.get());
        ScopedX509StoreCtx vctx(X509_STORE_CTX_new());
        X509_STORE_CTX_init(vctx.get(), store.get(), leaf_cert.get(), nullptr);
        int vret = X509_verify_cert(vctx.get());
        if (vret != 1) throw std::runtime_error("Certificate verification failed!");
        std::cout << "    [PASS] X.509 Certificate Chain signed by ML-DSA-65 Root CA verified!\n";

        // 3. Sender extracts ML-KEM-768 public key from verified cert and encapsulates
        std::cout << "[+] Step 3: Extracting ML-KEM-768 public key from verified certificate...\n";
        EVP_PKEY* extracted_pkey = X509_get0_pubkey(leaf_cert.get());
        if (!extracted_pkey) throw std::runtime_error("Failed to extract public key");

        std::cout << "[+] Step 4: Sender executing ML-KEM-768 Encap(pkR)...\n";
        ScopedPKEY_CTX ectx(EVP_PKEY_CTX_new(extracted_pkey, nullptr));
        EVP_PKEY_encapsulate_init(ectx.get(), nullptr);
        size_t ct_len = 0, secret_len = 0;
        EVP_PKEY_encapsulate(ectx.get(), nullptr, &ct_len, nullptr, &secret_len);
        std::vector<uint8_t> ct(ct_len), sender_secret(secret_len);
        EVP_PKEY_encapsulate(ectx.get(), ct.data(), &ct_len, sender_secret.data(), &secret_len);
        std::cout << "    Encapsulated Ciphertext size: " << ct.size() << " bytes (Expect 1088)\n";
        std::cout << "    Sender Shared Secret size   : " << sender_secret.size() << " bytes (32 bytes)\n";

        // 4. Sender encrypts payload with AES-256-GCM
        std::cout << "[+] Step 5: Sender encrypting confidential payload with AES-256-GCM...\n";
        std::string plaintext = "Authenticated End-to-End PQC Message delivered via X.509 PKI.";
        std::vector<uint8_t> nonce = {1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12};
        std::string aad = "PQC-PKI-v1-Auth";
        auto [ct_payload, tag] = aes_gcm_encrypt(sender_secret, nonce, plaintext, aad);

        // 5. Receiver loads private key and performs Decap
        std::cout << "[+] Step 6: Receiver decapsulating shared secret with private key...\n";
        FILE* key_f = fopen((pki_dir + "/receiver.key").c_str(), "r");
        if (!key_f) throw std::runtime_error("Failed to open receiver.key");
        ScopedPKEY rec_priv(PEM_read_PrivateKey(key_f, nullptr, nullptr, nullptr));
        fclose(key_f);

        ScopedPKEY_CTX dctx(EVP_PKEY_CTX_new(rec_priv.get(), nullptr));
        EVP_PKEY_decapsulate_init(dctx.get(), nullptr);
        std::vector<uint8_t> receiver_secret(secret_len);
        EVP_PKEY_decapsulate(dctx.get(), receiver_secret.data(), &secret_len, ct.data(), ct.size());

        if (sender_secret != receiver_secret) throw std::runtime_error("Decapsulated secret mismatch!");
        std::cout << "    [PASS] Decapsulated shared secret matches sender secret 100%!\n";

        // 6. Receiver decrypts payload
        std::cout << "[+] Step 7: Receiver decrypting payload and verifying AAD...\n";
        std::string decrypted = aes_gcm_decrypt(receiver_secret, nonce, ct_payload, tag, aad);
        assert(decrypted == plaintext);

        std::cout << "======================================================\n";
        std::cout << " [PASS] Decrypted message: \"" << decrypted << "\"\n";
        std::cout << " [SUCCESS] C++20 PQC X.509 PKI End-to-End verified!\n";
        std::cout << "======================================================\n";
        return 0;
    } catch (const std::exception& ex) {
        std::cerr << "[-] Error: " << ex.what() << "\n";
        return 1;
    }
}

File: examples/04-pki-and-x509/rust/src/main.rs
// =============================================================================
// Applied PQC Lab - PQC X.509 PKI End-to-End Encryption in Rust
// (ML-DSA-65 Root CA + ML-KEM-768 Leaf Cert + AES-256-GCM)
// =============================================================================

use foreign_types::ForeignType;
use openssl::pkey::PKey;
use openssl::stack::Stack;
use openssl::symm::{decrypt_aead, encrypt_aead, Cipher};
use openssl::x509::{store::X509StoreBuilder, X509StoreContext, X509};
use openssl_sys::*;
use std::env;
use std::fs;
use std::ptr;

fn main() -> Result<(), Box<dyn std::error::Error>> {
    let args: Vec<String> = env::args().collect();
    let pki_dir = if args.len() > 1 { &args[1] } else { "/tmp/pqc_pki_out" };

    println!("======================================================");
    println!(" [Applied PQC Lab] PQC X.509 PKI End-to-End in Rust");
    println!(" (ML-DSA-65 Root CA + ML-KEM-768 Leaf + AES-256-GCM)");
    println!("======================================================");

    // 1. Load Root CA and Leaf Certificates
    println!("[+] Step 1: Loading Root CA (ML-DSA-65) and Receiver Cert (ML-KEM-768)...");
    let ca_pem = fs::read(format!("{}/ca.crt", pki_dir))?;
    let ca_cert = X509::from_pem(&ca_pem)?;

    let leaf_pem = fs::read(format!("{}/receiver.crt", pki_dir))?;
    let leaf_cert = X509::from_pem(&leaf_pem)?;

    // 2. Sender verifies Certificate Chain
    println!("[+] Step 2: Sender verifying Receiver X.509 Certificate Chain...");
    let mut store_builder = X509StoreBuilder::new()?;
    store_builder.add_cert(ca_cert)?;
    let store = store_builder.build();
    let mut ctx = X509StoreContext::new()?;
    let chain = Stack::new()?;
    let is_valid = ctx.init(&store, &leaf_cert, &chain, |c| c.verify_cert())?;
    assert!(is_valid, "Certificate verification failed!");
    println!("    [PASS] X.509 Certificate Chain signed by ML-DSA-65 Root CA verified!");

    // 3. Sender extracts ML-KEM-768 public key and performs Encap
    println!("[+] Step 3: Extracting ML-KEM-768 public key from verified certificate...");
    let pubkey = leaf_cert.public_key()?;
    let pkey_ptr = pubkey.as_ptr();

    println!("[+] Step 4: Sender executing ML-KEM-768 Encap(pkR)...");
    let (ciphertext, sender_secret) = unsafe {
        let ectx = EVP_PKEY_CTX_new(pkey_ptr, ptr::null_mut());
        assert!(!ectx.is_null());
        assert!(EVP_PKEY_encapsulate_init(ectx, ptr::null_mut()) > 0);
        let mut ct_len = 0usize;
        let mut secret_len = 0usize;
        EVP_PKEY_encapsulate(ectx, ptr::null_mut(), &mut ct_len, ptr::null_mut(), &mut secret_len);
        let mut ct = vec![0u8; ct_len];
        let mut secret = vec![0u8; secret_len];
        assert!(EVP_PKEY_encapsulate(ectx, ct.as_mut_ptr(), &mut ct_len, secret.as_mut_ptr(), &mut secret_len) > 0);
        EVP_PKEY_CTX_free(ectx);
        (ct, secret)
    };
    println!("    Encapsulated Ciphertext size: {} bytes (Expect 1088)", ciphertext.len());
    println!("    Sender Shared Secret size   : {} bytes (32 bytes)", sender_secret.len());

    // 4. Sender encrypts payload with AES-256-GCM
    println!("[+] Step 5: Sender encrypting confidential payload with AES-256-GCM...");
    let plaintext = b"Authenticated End-to-End PQC Message delivered via X.509 PKI.";
    let nonce = [0x42u8; 12];
    let aad = b"PQC-PKI-v1-Auth";
    let mut tag = [0u8; 16];
    let ct_payload = encrypt_aead(Cipher::aes_256_gcm(), &sender_secret, Some(&nonce), aad, plaintext, &mut tag)?;

    // 5. Receiver loads private key and performs Decap
    println!("[+] Step 6: Receiver decapsulating shared secret with private key...");
    let rec_priv_pem = fs::read(format!("{}/receiver.key", pki_dir))?;
    let rec_priv = PKey::private_key_from_pem(&rec_priv_pem)?;

    let receiver_secret = unsafe {
        let dctx = EVP_PKEY_CTX_new(rec_priv.as_ptr(), ptr::null_mut());
        assert!(!dctx.is_null());
        assert!(EVP_PKEY_decapsulate_init(dctx, ptr::null_mut()) > 0);
        let mut secret = vec![0u8; sender_secret.len()];
        let mut secret_len = sender_secret.len();
        assert!(EVP_PKEY_decapsulate(dctx, secret.as_mut_ptr(), &mut secret_len, ciphertext.as_ptr(), ciphertext.len()) > 0);
        EVP_PKEY_CTX_free(dctx);
        secret
    };

    assert_eq!(sender_secret, receiver_secret);
    println!("    [PASS] Decapsulated shared secret matches sender secret 100%!");

    // 6. Receiver decrypts payload
    println!("[+] Step 7: Receiver decrypting payload and verifying AAD...");
    let decrypted = decrypt_aead(Cipher::aes_256_gcm(), &receiver_secret, Some(&nonce), aad, &ct_payload, &tag)?;
    assert_eq!(decrypted, plaintext);

    println!("======================================================");
    println!(" [PASS] Decrypted message: \"{}\"", String::from_utf8_lossy(&decrypted));
    println!(" [SUCCESS] Rust PQC X.509 PKI End-to-End verified!");
    println!("======================================================");

    Ok(())
}
